internal/management/controller/instance_controller.go

```diff
@@ -891,10 +893,28 @@ func (r *InstanceReconciler) RefreshSecrets(
 
 // reconcileInstance sets PostgreSQL instance parameters to current values
 func (r *InstanceReconciler) reconcileInstance(cluster *apiv1.Cluster) {
+	detectRequiresDesignatedPrimaryTransition := func() bool {
```

Comment:
This function is encapsulated here because this is the only place where the calculation should happen, and calling it elsewhere could lead to non-deterministic results and potential disasters